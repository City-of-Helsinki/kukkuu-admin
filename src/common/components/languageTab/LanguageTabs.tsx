import * as React from 'react';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import { useTranslate } from 'react-admin';

import { Language } from '../../../api/generatedTypes/globalTypes';
import styles from './languageTabs.module.css';

/**
 * Use as a wrapper for translatable content view
 * as render props, (please forgive me, HOC is worse).
 *
 * Example usage:
 * 1. Create local state using useState hook or component state
 * 2. import <LanguageTabs>, pass the required props, remember to use language enum.
 * 3. Basically render language tab and trigger onChange props that passed in
 * Gonna save a bit of boilterplate code.
 * Unfortunately HOC can not be done cause React-admin only pass props down to 1 level deep of children.
 *
 */
const LanguageTabs: React.FunctionComponent<{
  selectedLanguage: Language;
  onSelect: (value: Language) => void;
}> = ({ selectedLanguage, onSelect }) => {
  const translate = useTranslate();

  return (
    <Tabs
      indicatorColor="secondary"
      value={selectedLanguage}
      onChange={(e, value) => onSelect(value)}
      className={styles.languageTabs}
    >
      <Tab value={Language.FI} label={translate('languages.FI')} />
      <Tab value={Language.SV} label={translate('languages.SV')} />
      <Tab value={Language.EN} label={translate('languages.EN')} />
    </Tabs>
  );
};

export default LanguageTabs;
