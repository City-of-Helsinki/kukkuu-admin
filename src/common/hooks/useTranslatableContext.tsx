import { useContext } from 'react';

import TranslatableContext from '../contexts/TranslatableContext';

export default function useTranslatableContext() {
  return useContext(TranslatableContext);
}
