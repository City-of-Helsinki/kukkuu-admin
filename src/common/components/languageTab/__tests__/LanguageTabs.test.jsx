import * as React from 'react';
import { vi } from 'vitest';
import { render, fireEvent, screen } from '@testing-library/react';
import { useTranslate } from 'react-admin';

import LanguageTabs from '../LanguageTabs';
import useTranslatableContext from '../../../hooks/useTranslatableContext';

vi.mock('react-admin', () => ({
  useTranslate: vi.fn(),
}));

vi.mock('../../../hooks/useTranslatableContext', () => ({
  __esModule: true,
  default: vi.fn(),
}));

describe('LanguageTabs', () => {
  const mockTranslate = vi.fn();
  const mockSelectLanguage = vi.fn();
  const mockLanguages = ['en', 'fi', 'sv'];
  const mockSelectedLanguage = 'en';

  beforeEach(() => {
    useTranslate.mockReturnValue(mockTranslate);
    useTranslatableContext.mockReturnValue({
      languages: mockLanguages,
      selectLanguage: mockSelectLanguage,
      selectedLanguage: mockSelectedLanguage,
    });
    mockTranslate.mockImplementation((key) => key);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly', () => {
    const { container } = render(<LanguageTabs />);
    expect(container).toMatchSnapshot();
  });

  it('renders the correct number of tabs', () => {
    render(<LanguageTabs />);
    const tabs = screen.getAllByRole('tab');
    expect(tabs.length).toBe(mockLanguages.length);
  });

  it('calls selectLanguage when a tab is clicked', () => {
    render(<LanguageTabs />);
    const tab = screen.getByText('languages.fi');
    fireEvent.click(tab);
    expect(mockSelectLanguage).toHaveBeenCalledWith('fi');
  });
});
