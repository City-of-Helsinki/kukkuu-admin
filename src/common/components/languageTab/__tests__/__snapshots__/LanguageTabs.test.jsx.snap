// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LanguageTabs > renders correctly 1`] = `
<div>
  <div
    class="MuiTabs-root _languageTabs_1070fe css-1ujnqem-MuiTabs-root"
  >
    <div
      class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
      style="overflow: hidden; margin-bottom: 0px;"
    >
      <div
        class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
        role="tablist"
      >
        <button
          aria-selected="true"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          role="tab"
          tabindex="0"
          type="button"
        >
          languages.en
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          role="tab"
          tabindex="-1"
          type="button"
        >
          languages.fi
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          role="tab"
          tabindex="-1"
          type="button"
        >
          languages.sv
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <span
        class="MuiTabs-indicator css-11yukd5-MuiTabs-indicator"
        style="left: 0px; width: 0px;"
      />
    </div>
  </div>
</div>
`;
